File: scripts/DimmerPublisher.py
#!/usr/bin/python
import rospy
from openhab_msgs.msg import DimmerCommand

class DimmerPublisher(object):
    """Node example class."""

    def __init__(self, item_name):

        self.item_name = item_name
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", DimmerCommand, queue_size=10)
        self.rate = rospy.Rate(10) # 10hz

        # Initialize message variables.
        self.enable = False
        self.message = None

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        """Turn on publisher."""
        self.enable = True
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", DimmerCommand, queue_size=10)

        while not rospy.is_shutdown():

            self.message = DimmerCommand()

            self.message.command = DimmerCommand.ON
            self.message.iscommand = False
            self.message.ispercentage = True
            self.message.percentage = DimmerCommand.PERCENTAGE80
            
            self.message.isnull = False

            self.message.header.stamp = rospy.Time.now()
            self.message.header.frame_id = "/base_link"
            self.message.item = self.item_name

            message = "Publishing to %s at %s: command = %s, iscommand = %s, ispercentage = %s, percentage = %s" % (self.message.item, rospy.get_time(), self.message.command, self.message.iscommand, self.message.ispercentage, self.message.percentage)
            rospy.loginfo(message)

            self.pub.publish(self.message)
            self.rate.sleep()

    def stop(self):
        """Turn off publisher."""
        self.enable = False
        self.pub.unregister()

# Main function.
if __name__ == "__main__":
    # Initialize the node and name it.
    rospy.init_node("DimmerPublisherNode", anonymous=True)
    # Go to class functions that do all the heavy lifting.

    dimmerPublisher = DimmerPublisher("testDimmer")

    try:
        dimmerPublisher.start()
    except rospy.ROSInterruptException:
        pass
    # Allow ROS to go to all callbacks.
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()

File: src/DimmerPublisher.cpp
#include <ros/ros.h>
#include "openhab_msgs/DimmerCommand.h"

class DimmerPublisher
{
public:
    // Constructor
    DimmerPublisher(const std::string& item_name)
        : item_name_(item_name), enable_(false)
    {
        pub_ = nh_.advertise<openhab_msgs::DimmerCommand>("/openhab/items/" + item_name_ + "/command", 10);
        rate_ = ros::Rate(10); // 10hz
        stop();
    }

    void start()
    {
        // Turn on publisher.
        enable_ = true;
        pub_ = nh_.advertise<openhab_msgs::DimmerCommand>("/openhab/items/" + item_name_ + "/command", 10);

        while (ros::ok() && enable_)
        {
            openhab_msgs::DimmerCommand message;

            message.command = openhab_msgs::DimmerCommand::ON;
            message.iscommand = false;
            message.ispercentage = true;
            message.percentage = openhab_msgs::DimmerCommand::PERCENTAGE80;

            message.isnull = false;

            message.header.stamp = ros::Time::now();
            message.header.frame_id = "/base_link";
            message.item = item_name_;

            std::stringstream ss;
            ss << "Publishing to " << message.item << " at " << ros::Time::now() << ": command = " << message.command
               << ", iscommand = " << message.iscommand << ", ispercentage = " << message.ispercentage
               << ", percentage = " << message.percentage;
            ROS_INFO_STREAM(ss.str());

            pub_.publish(message);
            rate_.sleep();
        }
    }

    void stop()
    {
        // Turn off publisher.
        enable_ = false;
        pub_.shutdown();
    }

private:
    std::string item_name_;
    bool enable_;
    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Rate rate_;
};

int main(int argc, char** argv)
{
    // Initialize the node and name it.
    ros::init(argc, argv, "DimmerPublisherNode", ros::init_options::AnonymousName);

    // Go to class functions that do all the heavy lifting.
    DimmerPublisher dimmer_publisher("testDimmer");

    try
    {
        dimmer_publisher.start();
    }
    catch (const ros::Exception& e)
    {
        ROS_ERROR_STREAM("ROS Exception: " << e.what());
    }

    // Allow ROS to go to all callbacks.
    // spin() simply keeps C++ from exiting until this node is stopped.
    ros::spin();

    return 0;
}
